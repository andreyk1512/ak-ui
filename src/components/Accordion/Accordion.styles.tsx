import { css } from '@emotion/react';
import * as styles from '../../styles';

export const containerStyles = css`
  border-radius: 6px;
  overflow: hidden;
  box-sizing: border-box;
  background-color: ${styles.color.white};
  border: 1px solid ${styles.color.brightGray_70};
`;

export const headerStyles = css`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 25px 20px;
  box-sizing: border-box;
  cursor: pointer;

  h3 {
    margin: 0;
  }
`;

export const iconStyles = (expanded: boolean) => css`
  margin: -6px;
  transition: transform linear 0.2s;
  transform: rotate(0turn);

  & {
    circle {
      transition: fill linear 0.2s;
    }
    path {
      transition: fill linear 0.2s;
    }
  }

  ${expanded &&
  css`
    & {
      transform: rotate(0.5turn);

      circle {
        fill: ${styles.color.lightSeaGreen};
      }
      path {
        fill: ${styles.color.white};
      }
    }
  `}
`;

export const bodyStyles = (expanded: boolean) => css`
  visibility: hidden;
  overflow: hidden;
  max-height: 0;
  transition: all 0.2s linear;
  border-top: 0 solid ${styles.color.brightGray_85};
  box-sizing: border-box;

  ${expanded &&
  css`
    visibility: visible;
    border-top-width: 1px;
  `}
`;

export const bodyInnerStyles = css`
  margin: 20px;
`;
