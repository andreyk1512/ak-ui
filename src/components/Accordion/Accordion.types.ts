import { ReactNode } from 'react';

export type AccordionProps = {
  id?: string;
  title: string;
  children: ReactNode;
  defaultExpanded?: boolean;
};
