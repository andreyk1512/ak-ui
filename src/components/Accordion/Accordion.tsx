import { useEffect, useRef, useState } from 'react';
import { ReactComponent as AccordionArrowIcon } from '../../assets/icons/accordion-arrow.svg';
import { Typography } from '../Typography';
import { bodyInnerStyles, bodyStyles, containerStyles, headerStyles, iconStyles } from './Accordion.styles';
import { AccordionProps } from './Accordion.types';

export const Accordion = ({ id, title, children, defaultExpanded = false }: AccordionProps) => {
  const ref = useRef<HTMLDivElement>(null);
  const [expanded, setExpanded] = useState(defaultExpanded);

  useEffect(() => {
    if (ref.current) {
      if (expanded) {
        ref.current.style.maxHeight = `${ref.current.scrollHeight}px`;
      } else {
        ref.current.style.maxHeight = '0';
      }
    }
  }, [expanded]);

  const switchExpand = () => setExpanded((prevState) => !prevState);

  return (
    <div css={containerStyles} id={id}>
      <div css={headerStyles} onClick={switchExpand}>
        <Typography.Title type={3}>{title}</Typography.Title>
        <AccordionArrowIcon css={iconStyles(expanded)} title="expand" />
      </div>
      <div css={bodyStyles(expanded)} ref={ref}>
        <div css={bodyInnerStyles}>{children}</div>
      </div>
    </div>
  );
};
