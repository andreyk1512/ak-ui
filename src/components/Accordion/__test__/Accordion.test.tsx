import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { Accordion } from '../Accordion';

describe('Accordion', () => {
  it('should render the correct snapshot of the collapsed state', () => {
    const { container } = render(<Accordion title="My Accordion">Children</Accordion>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the correct snapshot of the expanded state', () => {
    const { container } = render(<Accordion title="My Accordion">Children</Accordion>);
    fireEvent.click(screen.getByText('My Accordion'));

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with expanded state if defaultExpanded prop is true', () => {
    render(<Accordion title="My Accordion" defaultExpanded>Children</Accordion>);
    const body = screen.getByText('Children');

    expect(body).toBeVisible();
  });

  it('should render the title', () => {
    const title = 'My Accordion';
    render(<Accordion title={title}>Children</Accordion>);
    const titleElement = screen.getByText(title);

    expect(titleElement).toBeInTheDocument();
  });

  it('should render the arrow icon', () => {
    render(<Accordion title="My Accordion">Children</Accordion>);
    const icon = screen.getByTitle('expand');

    expect(icon).toBeInTheDocument();
  });

  it('should collapse the body by default', () => {
    render(<Accordion title="My Accordion">Children</Accordion>);
    const body = screen.getByText('Children');

    expect(body).not.toBeVisible();
  });

  it('should expand the body when clicked', () => {
    render(<Accordion title="My Accordion">Children</Accordion>);
    const header = screen.getByText('My Accordion');
    fireEvent.click(header);
    const body = screen.getByText('Children');

    expect(body).toBeVisible();
  });

  it('should collapse the body when clicked again', () => {
    render(<Accordion title="My Accordion">Children</Accordion>);
    const header = screen.getByText('My Accordion');
    fireEvent.click(header);
    fireEvent.click(header);
    const body = screen.getByText('Children');

    expect(body).not.toBeVisible();
  });
});
