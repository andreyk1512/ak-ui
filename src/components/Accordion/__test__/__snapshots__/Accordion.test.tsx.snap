// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render the correct snapshot of the collapsed state 1`] = `
<div
  class="css-16ztqdi"
>
  <div
    class="css-13khp69"
  >
    <h3
      class="css-wcntdc"
    >
      My Accordion
    </h3>
    <span
      class="css-1lrz23l"
      title="expand"
    />
  </div>
  <div
    class="css-1mdb6nq"
    style="max-height: 0;"
  >
    <div
      class="css-1frs77m"
    >
      Children
    </div>
  </div>
</div>
`;

exports[`Accordion should render the correct snapshot of the expanded state 1`] = `
<div
  class="css-16ztqdi"
>
  <div
    class="css-13khp69"
  >
    <h3
      class="css-wcntdc"
    >
      My Accordion
    </h3>
    <span
      class="css-6hkuxv"
      title="expand"
    />
  </div>
  <div
    class="css-lf2lug"
    style="max-height: 0px;"
  >
    <div
      class="css-1frs77m"
    >
      Children
    </div>
  </div>
</div>
`;
